hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/ChunkManagerImpl.java

```diff
@@ -210,19 +216,39 @@ public ByteBuffer readChunk(Container container, BlockID blockID,
     // of the chunk file.
     if (containerData.getLayOutVersion() == ChunkLayOutVersion
         .getLatestVersion().getVersion()) {
-      File chunkFile = ChunkUtils.getChunkFile(containerData, info);
 
-      // In case the chunk file does not exist but tmp chunk file exist,
-      // read from tmp chunk file if readFromTmpFile is set to true
-      if (!chunkFile.exists() && dispatcherContext != null
-          && dispatcherContext.isReadFromTmpFile()) {
-        chunkFile = getTmpChunkFile(chunkFile, dispatcherContext);
+      File finalChunkFile = ChunkUtils.getChunkFile(containerData, info);
+      File tmpChunkFile = getTmpChunkFile(finalChunkFile, dispatcherContext);
+
+      List<File> possibleFiles = new ArrayList<>();
+      possibleFiles.add(finalChunkFile);
+      if (dispatcherContext.isReadFromTmpFile()) {
+        possibleFiles.add(tmpChunkFile);
+        possibleFiles.add(finalChunkFile);
       }
-      data = ChunkUtils.readData(chunkFile, info, volumeIOStats);
-      containerData.incrReadCount();
-      long length = chunkFile.length();
-      containerData.incrReadBytes(length);
-      return data;
+
+      boolean found = false;
+
+      for (File chunkFile : possibleFiles) {
+        try {
+          data = ChunkUtils.readData(chunkFile, info, volumeIOStats);
+          containerData.incrReadCount();
+          long length = chunkFile.length();
```

Comment:
```suggestion
          long length = info.getLen();
```

Rationale:

1. avoid another race condition (`chunkFile` being gone by this time)
2. file might be larger than length of data read (since `info` may specify offset), though I think currently offset is always 0

Reply:
Thanks, updated.

(BTW, I don't know why this `incrReadBytes` is used at all as we already have `volumeIOStats.incReadBytes` in ChunkUtils. But that's an other question.)